integration/smartbft/mock_orderer.go

```diff
@@ -184,19 +180,25 @@ func (mo *MockOrderer) deliverBlocks(
 			return status, nil
 		}
 
-		// increment block number to support FAIL_IF_NOT_READY deliver behavior
 		number++
-		mo.logger.Infof("[channel: %s] Delivering block [%d] for (%p) for %s", chdr.ChannelId, block.Header.Number, seekInfo, addr)
 
 		block2send := &cb.Block{
 			Header:   block.Header,
 			Metadata: block.Metadata,
 			Data:     block.Data,
 		}
+
 		if seekInfo.ContentType == ab.SeekInfo_HEADER_WITH_SIG && !protoutil.IsConfigBlock(block) {
-			mo.logger.Infof("ASKED FOR HEADER WITH SIG")
 			block2send.Data = nil
+		} else if mo.censorDataMode {
+			if !mo.peerFirstTry {
+				mo.malicious = true
+			}
+			if mo.sentCount >= 5 && mo.malicious {
+				break
+			}
 		}
+		mo.peerFirstTry = true
```

Comment:
This style of controlling the mock will only cover one use case.
We want to achieve the same goal with a more flexible API on the mock, not by setting a single private field `mo.censorDataMode`.

I am thinking about the following:

on every mock you can have the following functions:

// will deliver either data/headers until this point "n" and will then wait for further instructions but NOT break the delivery loop. If a subsequent call DeliverUntil(n2 uint64) (n2>n) the mock will continue to deliver until n2.
DeliverUntil(number uint64)

// answers true if the mock was last requested to deliver data
IsDataDeliverer() bool

// answers true if the mock was last requested to deliver headers
IsHeaderDeliverer() bool

With this, the "block censorship" test case can then be implemented by:

1. mark all mocks by DeliverUntil(5)
2. discover who IsDataDeliverer and who are IsHeaderDeliverer (with eventually)
3. let the header deliverers continue until the end by marking them DeliverUntil(11)
4. etc...

I hope you see how this style will allow us to test many more test cases with ease.

The "normal" case is just
1.  mark all mocks by DeliverUntil(11)
2. etc...

Reply:
https://github.com/hyperledger/fabric/issues/4724